### ISSUE TITLE
Should allow pressure-vessel to make /var/lib/snapd/desktop visible to its container

### ISSUE THREAD

**Issue by smcv** [canonical/steam-snap#369]
### Ensure there isn't an existing issue for this and check the wiki

- [X] This issue is not a duplicate and I have checked the wiki.

### Current Behavior

While testing a new version of Steam Linux Runtime 3.0 (sniper) I encountered this error:

```
bwrap: Can't bind mount /oldroot/var/lib/snapd/desktop on /newroot/var/lib/snapd/desktop: Unable to mount source on destination: Permission denied
```

This is because the new version of SLR specifically makes the following paths available to the container, in order to avoid frequent crashes when users do inadvisable things with symlinks:

* `~/.cache`
* `~/.config` (this is the one that is actually a functional requirement, for [steam-for-linux#10547](https://github.com/ValveSoftware/steam-for-linux/issues/10547), [steam-for-linux#10548](https://github.com/ValveSoftware/steam-for-linux/issues/10548), [steam-for-linux#10552](https://github.com/ValveSoftware/steam-for-linux/issues/10552), [steam-for-linux#10558](https://github.com/ValveSoftware/steam-for-linux/issues/10558), [steam-for-linux#10560](https://github.com/ValveSoftware/steam-for-linux/issues/10560))
* `~/.local/share`
* `$XDG_CACHE_HOME`
* `$XDG_CONFIG_HOME`
* `$XDG_CONFIG_DIRS`, excluding paths below `/usr` if any
* `$XDG_DATA_HOME`
* `$XDG_DATA_DIRS`, excluding paths below `/usr` if any (this is the one that was causing the crash for me)

### Expected Behavior

Steam works. SLR can bind-mount everything it wants to bind-mount, notably `/var/lib/snapd/desktop` and any other paths that are listed in important environment variables.

### Steps To Reproduce

You can probably emulate the behaviour of the unreleased SLR with released versions by running with `PRESSURE_VESSEL_FILESYSTEMS_RW=/var/lib/snapd/desktop` in the environment.

This is untested: sorry, I do not currently have the time available to carry out full debugging on a Snap app that I don't maintain. I wanted to open this issue to give the maintainers of this app a heads-up that new Steam releases will probably regress very soon if the AppArmor profile is not made less restrictive.

### Environment

```yaml
not provided
```

### gaming-graphics-core22 version

kisak-fresh (default)

**Comment by smcv** [comment]
Note that as of last week, the ability to run Steam Linux Runtime 3.0 (sniper) is a hard dependency for Steam's UI, so anything that breaks SLR is now a critical issue that will prevent users from running Steam.

The SLR version that I am currently testing is required right now to avoid common crashes seen by non-Snap users, so it is likely that I will be asked to release it immediately, even if it breaks Snap as a side-effect; I cannot necessarily afford the time needed to add extra workarounds for Snap right now.

**Comment by smcv** [comment]
> The SLR version that I am currently testing is required right now to avoid common crashes seen by non-Snap users, so it is likely that I will be asked to release it immediately, even if it breaks Snap as a side-effect

This is now very likely to be included in the next Steam release.

**Comment by ZoopOTheGoop** [comment]
Two important questions:

1.

When you say `~` how are you constructing this? Literally `~`, `~$USER`, `$HOME`, or an explicit `/home/$USER`? This has become relevant in snap contexts before. (We rebind things so that `~` and `$HOME` are rebound in the snap environment, but `$USER` is still the host user's name, so making sure).

2.

Will SLR crash if *any* value under `$XDG_DATA_DIRS`  or `$XDG_CONFIG_DIRS` (nor under `/usr`) is not rwbindable? 

I ask, because currently the snap rebinds the other `$XDG_*` variables to its own locations, but those two actually prepend new values to the host's settings. For instance, my `$XDG_DATA_DIRS` contains (among other things) `/home/<me>/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share` in addition to `/var/lib/snapd/desktop`, and my `$XDG_CONFIG_DIRS` includes  (among other things) `/home/<me>/.config/kdedefaults:/etc/xdg`. 

Of those, we can probably make a patch including `/etc/xdg` and maybe `/var/lib/snapd/desktop` approved, but everything else is more or less entirely dependent on the user's setup (for instance having plasma and flatpak installed), and anything under `$HOME` in going to be difficult to get approved (which isn't your problem, but I just want to say for my own benefit as I'm trying to fix this).

**Comment by ZoopOTheGoop** [comment]
As a note to self, the *best* route here, at least while getting a `snapd` patch approved, is probably to rebind those variables entirely in the desktop launch script to known good paths, or if being slightly fancier, iterate over the paths and remove any that aren't accessible by Steam.

**Comment by smcv** [comment]
> When you say `~` how are you constructing this? Literally `~`, `~$USER`, `$HOME`, or an explicit `/home/$USER`?

It's GLib's `g_get_home_dir()`, which uses `$HOME`, falling back to the "official" home directory from the equivalent of `getent passwd $USER` if `$HOME` is somehow unset. So in the Snap case, we'll be using the Snap-specific fake home, not the real home.

> Will SLR crash if _any_ value under `$XDG_DATA_DIRS` or `$XDG_CONFIG_DIRS` (nor under `/usr`) is not rwbindable?

More precisely, it is any directory in those paths that exists, but cannot be bind-mounted, and is not below a reserved directory (which in practice mainly means `/usr`). If the directory doesn't exist at all in the Snap environment, we'll automatically skip it.

If the directory exists but cannot be opened due to permissions or AppArmor rules, I'm not sure without checking the implementation what would happen.

There is a workaround queued up for a future Steam beta which hacks out the one specific directory `/var/lib/snapd/desktop` from the bubblewrap command-line, but that missed the boat for the hotfix that recently reached Steam betas, and in any case I'd prefer to be able to remove Snap-specific hacks like that to reduce maintenance overhead. You can test this workaround by replacing `pressure-vessel/` with the result of unpacking <https://repo.steampowered.com/pressure-vessel/snapshots/0.20240306.0/pressure-vessel-bin.tar.gz>.

> I ask, because currently the snap rebinds the other `$XDG_*` variables to its own locations, but those two actually prepend new values to the host's settings. For instance, my `$XDG_DATA_DIRS` contains (among other things) `/home/<me>/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share` in addition to `/var/lib/snapd/desktop`, and my `$XDG_CONFIG_DIRS` includes (among other things) `/home/<me>/.config/kdedefaults:/etc/xdg`. None of which are available to the Snap environment.

What does "not available" mean in this context? `ENOENT` or `EACCES` or what?

It isn't clear to me why these directories would be advertised to the Snap app as "you should read data/config from here" (by leaving them in the environment) if there is a security policy that says that actually, it should not? If the result is that the Snap app tries to access those directories, that seems to me rather like "you asked for it, you got it".

**Comment by smcv** [comment]
For context, the reason why SLR needs to do this is that it seems to be somewhat frequent for well-intentioned users to replace their `/home/me/.config` with a symlink to `/ssd/home/me/config`. This is fine (ish) when not in a container, but in a container environment where `$HOME` is bind-mounted but `/ssd` is not, it will often result in `/home/me/.config` being a dangling symlink. Empirically, the CEF framework used in `steamwebhelper` does not handle this at all gracefully, crashing with no error message.

If users are doing this for `~/.config` in the config stack, then they could equally well be doing this for `~/.local/share` in the data stack, and they could plausibly be running with something like `XDG_DATA_HOME=/home/me/.private`, `XDG_DATA_DIRS=/home/me/.local/share:/usr/local/share:/usr/share` - which is why we currently do this symmetrically for `XDG_DATA_HOME` and all elements of `XDG_DATA_DIRS` that are not in `/usr`, as well as the corresponding config stack.

**Comment by ZoopOTheGoop** [comment]
> What does "not available" mean in this context? ENOENT or EACCES or what?
>
> It isn't clear to me why these directories would be advertised to the
Snap app as "you should read data/config from here" (by leaving them in the
environment) if there is a security policy that says that actually, it
should not? If the result is that the Snap app tries to access those
directories, that seems to me rather like "you asked for it, you got it".

Any sense, basically the issue here is that snaps mirror the environment
variables of the host. This is useful for, say, passing environment
variables directly to the program running in confinement but less useful
for anything else. There are a few paths that are rebound, but they're very
limited (stuff like `$PATH` and `$HOME`). The "workaround" for this is that
the desktop SDK extensions (gnome and kde) inject launch scripts that run
before the wrapped program and mess with the environment variables.

Steam uses basically a fork of the Gnome launch script, (which I'm finding
out on my recent installation has... interesting effects if you are on a
pure KDE environment), and for whatever reason THAT script (or at least the
way it was at the time it was copied - before my time) just prepends
directories to the user's `XDG_*_DIRS` variables (while rebinding the
single directory ones like `XDG_CONFIG`).

Why? To be honest, I have no idea, in practice these directories are
unlikely to be accessible to the snap, given that *any* random directory is
likely unavailable to the snap environment without specific effort to make
it available. I suppose the reasoning was probably that any application
would find what it needed in the snap directories and then not check the
User ones, and if it did, sometimes it would get lucky and find an
accessible one while gracefully ignoring the inaccessible ones. (This is
admittedly occasionally useful, it does let us give the Steam Snap access
to the host font config, which lets it use MS fonts when available on the
host system instead of being stuck with Liberation).

We only really touch that script when we have to, there's a lot of code
that I suspect is dead but can't be 100% sure of, and we only really go in
and mess with it when we start seeing AppArmor violations that cause actual
bugs. It's not a... good state of affairs, but it is the explanation.
Essentially "we use a modified launch script originally meant for a
different environment that nobody currently working on the Snap wrote that
sometimes makes problematic decisions we haven't changed because nothing
has happened yet that even has made us aware of the behavior."

**Comment by ZoopOTheGoop** [comment]
That said, this isn't a problem entirely unique to the Snap environment,
just exacerbated by it, a sufficiently haphazard user could easily alter
their variables to point to nonsense directories they can't access.

**Comment by smcv** [comment]
pressure-vessel 0.20240306.0, as included in Steam client update 1709846872, has a temporary workaround for this. The workaround will be removed in a future release.

To reproduce the issue, replace the `pressure-vessel/` directory with the result of unpacking https://repo.steampowered.com/pressure-vessel/snapshots/0.20240301.0/pressure-vessel-bin.tar.gz which does not have this workaround. (In future pressure-vessel releases I'll probably add an environment variable that turns off workarounds, to facilitate testing of this sort of thing.)

**Comment by ashuntu** [comment]
A partial workaround for this is now in the edge channel of the Steam snap, which we will try to get promoted up the channels by the end of the week if not sooner. We also intend on fixing this on the snapd side for a more permanent solution.

**Comment by thatLeaflet** [comment]
> the ability to run Steam Linux Runtime 3.0 (sniper) is a hard dependency for Steam's UI, so anything that breaks SLR is now a critical issue that will prevent users from running Steam

Does this mean that the Steam client will use libraries from the Sniper runtime to run? My current understanding is that the client uses the Scout runtime (based on Ubuntu 12.04 LTS libraries) to run.

**Comment by smcv** [comment]
If I'm keeping track correctly, this was fixed in snapd 2.62, which is mandatory since #367. As a result, the pressure-vessel version included in yesterday's SLR soldier and sniper beta releases disables our workaround for this issue by default. If there are no regression reports in the next few days/weeks, we will remove the workaround completely.

The status of this issue is similar to https://github.com/canonical/steam-snap/issues/356#issuecomment-1912411145.

If it becomes necessary to re-enable the workaround, you can do this:

1. Report an issue to https://github.com/ValveSoftware/steam-runtime/issues so we can enable it by default again, and so that we know that we can't delete the implementation of the workaround yet
2. Run with `PRESSURE_VESSEL_WORKAROUNDS="+steam-snap#369"` in the environment (if you need to use more than one workaround, it's a space-separated list)
